tests/README.md

```diff
@@ -0,0 +1,1510 @@
+			->withArgs([$resource, $length])
+			->andReturn(false);
+
+		$functionMock->shouldReceive('fclose')
+			->once()
+			->with($resource);
+
+		//set the exception assertion
+		$this->expectException(\RuntimeException::class);
+		$this->expectExceptionMessage('Unable to read from stream');
+
+		//act
+		$subject->read($length);
+	}
+```  
+The prefix *alias* before the class name tells mockery to load an empty class with the same name.
+Afterwards exceptions are placed on the empty mock alias. It is important that these tests must run in
+separate PHP processes because PHP has no functionality to unload classes. Therefore, a redefinition of
+a class would lead to a fatal error. PHPUnit has a build in function which does that. Every class which is
+annotated with *@runTestsInSeparateProcesses* will spawn a new PHP process for each unit test.
+
+<a name="fluent-interfaces"></a>
+#### Fluent interfaces
+Fluent interfaces or long call chains might be easy to read and leverage but kind of hard to test because
+a lot of expectation have to be set on many mocks which are not really interesting for the test at all.
+PHPUnit has a feature which simplifies this process a lot.
+
+For example file upload test have to mock the http service. The http service has methods like request and response and subsequent methods
+to interact with them. To avoid to create unneeded expectations on the http mock the request mock is generated on the fly
+with the expectation set on it. The pattern is *{method name}->{method name}* the last method gets the actual expectations defined after the
+*shouldReceive* call.
+
+```php
+<?php
+//class FileUploadImplTest
+	/**
+	 * @Test
+	 * @small
+	 */
+	public function testRegisterWithProcessedFilesWhichShouldFail() {
+		$processorMock = \Mockery::mock(PreProcessor::class);
+		//create a request mock on the fly and set an expectation on it with the arrows (->).
+		$this->globalHttpStateMock->shouldReceive('request->getUploadedFiles')
+			->once()
+			->andReturn([]);
+
+		$this->expectException(IllegalStateException::class);
+		$this->expectExceptionMessage('Can not register processor after the upload was processed.');
+
+		$this->subject->process();
+		$this->subject->register($processorMock);
+	}
+```
+
+<a name="rise-testability"></a>
+### Rise testability
+<a name="solid"></a>
+#### SOLID
+Robert C. Martin gathered 5 principles for object oriented class design, for building maintainable object oriented system (see Langr 2015).
+<a name="single-responsibility-principle-srp"></a>
+##### Single Responsibility Principle (SRP)
+Classes should have one reason to change. Keep the classes small and single-purposed.
+<a name="open-closed-principle-ocp"></a>
+##### Open-Closed Principle (OCP)
+Classes should be design to be open for extension but closed for modification. The need to make changes to existing classes
+should be minimized.
+<a name="liskov-substitution-principle-lsp"></a>
+##### Liskov Substitution Principle (LSP)
```

Comment:
This is especially hard, you might want to link the "colored point" or "shape" example here, that show how easily that principle is violated. The general take away from that should be: do not use inheritance.